src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,141 @@
+
     @Override
     public void add(T value) {
-
+        growIfNeeded();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        addElementArrayListIndexOutOfBoundException(index);
+        growIfNeeded();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        growIfNeeded(list);
+        int listSize = list.size();
+        for (int i = 0; i < listSize; i++) {
+            elements[size + i] = list.get(i);
+        }
+        size += listSize;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        arrayListIndexOutOfBoundException(index);
+        return (T)elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        arrayListIndexOutOfBoundException(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        arrayListIndexOutOfBoundException(index);
+        T removedElement = (T)elements[index];
+        removeElement(index);
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int indexRemovedElement = -1;
+        for (int i = 0; i < size; i++) {
+            if (elements[i] != null
+                    && elements[i].equals(element)
+                    || elements[i] == element) {
+                indexRemovedElement = i;
+                break;
+            }
+        }
+        if (indexRemovedElement == -1) {
+            throw new NoSuchElementException("Can't find element: " + element);
+        }
+        T removedElement = (T)elements[indexRemovedElement];
+        removeElement(indexRemovedElement);
+        return removedElement;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void growIfNeeded() {
+        if (elements.length == 1 && elements.length == size) {
+            Object[] newArray = new Object[DEFAULT_CAPACITY];
+            System.arraycopy(elements, 0, newArray, 0, size);
+            elements = newArray;
+        }
+        if (elements.length == size) {
+            Object[] newArray = new Object[(int)((double)elements.length * GROWTH_FACTOR)];
+            System.arraycopy(elements, 0, newArray, 0, size);
+            elements = newArray;
+        }
+    }
+
+    private void growIfNeeded(List<T> list) {
+        if (list != null && elements.length < size + list.size()) {
+            Object[] newArray = new Object[size + list.size()];
+            System.arraycopy(elements, 0, newArray, 0, size);
+            elements = newArray;
+        }
+    }
+
+    private void addElementArrayListIndexOutOfBoundException(int index) {
```

Comment:
bad name for a method, usually Exceptions end with the word Exception, not methods